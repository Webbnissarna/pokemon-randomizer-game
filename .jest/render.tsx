import { render as rtlRender } from "@testing-library/react";
import { InitializeColorMode, ThemeProvider } from "theme-ui";

import theme from "../utils/theme";

import "./style.css";

export function render(ui: React.ReactElement) {
  return rtlRender(
    <>
      <InitializeColorMode />
      <ThemeProvider theme={theme}>{ui}</ThemeProvider>
    </>
  );
}
