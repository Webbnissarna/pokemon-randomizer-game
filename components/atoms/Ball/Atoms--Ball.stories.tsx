import React from "react";
import { Flex } from "theme-ui";
import Ball from ".";
import type { BallProps, ClickEvent, Size } from "./Ball.types";

export function Default({ size, disabled }: BallProps): JSX.Element {
  const sizes: [string, Size][] = [
    ["Small", "sm"],
    ["Large", "lg"],
  ];

  const onClick = (event: ClickEvent) => {
    console.log("click", event);
  };

  return (
    <Flex sx={{ flexDirection: "column", gap: 20 }}>
      <h1>Custom</h1>
      <Ball size={size} disabled={disabled} onClick={onClick} />

      {sizes.map(([title, ballSize]) => (
        <React.Fragment key={title}>
          <h1>{title}</h1>
          <Ball size={ballSize} onClick={onClick} />
          <Ball size={ballSize} onClick={onClick} disabled={true} />
        </React.Fragment>
      ))}
    </Flex>
  );
}

Default.args = {
  size: "sm",
  disabled: false,
} as BallProps;

Default.argTypes = {
  size: {
    options: ["sm", "lg"],
    control: { type: "select" },
  },
};
