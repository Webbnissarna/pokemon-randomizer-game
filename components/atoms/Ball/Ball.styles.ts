import type { ThemeUIStyleObject } from "theme-ui";
import type { Size } from "./Ball.types";

export const containerStyle: ThemeUIStyleObject = {
  borderRadius: "50%",
  overflow: "hidden",
  transition: "transform 0.05s ease-out",
};

export const sizeMap: Record<Size, number> = {
  sm: 40,
  lg: 100,
};

export const activeStyle: ThemeUIStyleObject = {
  "&:hover": {
    cursor: "pointer",
  },

  "&:active": {
    transform: "scale(0.95)",
  },
};

export function getSizeStyle(size: Size): ThemeUIStyleObject {
  return {
    width: sizeMap[size],
    height: sizeMap[size],
  };
}
