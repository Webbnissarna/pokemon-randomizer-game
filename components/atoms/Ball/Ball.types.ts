import type { MouseEvent as ReactMouseEvent } from "react";

export type Size = "sm" | "lg";

export type ClickEvent = ReactMouseEvent<HTMLButtonElement, MouseEvent>;

export interface BallProps {
  size: Size;
  disabled?: boolean;
  onClick?: (event: ClickEvent) => void;
}
