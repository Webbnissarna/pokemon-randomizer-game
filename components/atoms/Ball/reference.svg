<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!-- Created with Inkscape (http://www.inkscape.org/) -->

<svg
   width="26.458336mm"
   height="26.458361mm"
   viewBox="0 0 26.458336 26.458361"
   version="1.1"
   id="svg5"
   sodipodi:docname="reference.svg"
   inkscape:version="1.1.2 (0a00cf5339, 2022-02-04, custom)"
   xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape"
   xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
   xmlns="http://www.w3.org/2000/svg"
   xmlns:svg="http://www.w3.org/2000/svg">
  <sodipodi:namedview
     id="namedview7"
     pagecolor="#888888"
     bordercolor="#666666"
     borderopacity="1.0"
     inkscape:pageshadow="2"
     inkscape:pageopacity="0"
     inkscape:pagecheckerboard="true"
     inkscape:document-units="mm"
     showgrid="false"
     inkscape:zoom="4"
     inkscape:cx="43.625"
     inkscape:cy="67.75"
     inkscape:window-width="2495"
     inkscape:window-height="1417"
     inkscape:window-x="-8"
     inkscape:window-y="-8"
     inkscape:window-maximized="1"
     inkscape:current-layer="layer1" />
  <defs
     id="defs2" />
  <g
     inkscape:label="Layer 1"
     inkscape:groupmode="layer"
     id="layer1"
     transform="translate(-91.770834,-141.88542)">
    <path
       id="path868"
       style="fill:#ff6060;fill-opacity:1;stroke-width:0.7;stroke-linecap:round;stroke-linejoin:round;stop-color:#000000"
       d="M 105,141.88542 A 13.229166,13.229166 0 0 0 91.770834,155.11458 H 118.22917 A 13.229166,13.229166 0 0 0 105,141.88542 Z" />
    <path
       id="path1155"
       style="fill:#ffffff;stroke-width:0.7;stroke-linecap:round;stroke-linejoin:round;stop-color:#000000"
       d="m 105,168.34378 a 13.229166,13.229166 0 0 0 13.22917,-13.22916 H 91.770834 A 13.229166,13.229166 0 0 0 105,168.34378 Z"
       inkscape:transform-center-y="6.5718179" />
    <rect
       style="fill:#606060;fill-opacity:1;stroke-width:0.699999;stroke-linecap:round;stroke-linejoin:round;stop-color:#000000"
       id="rect1467"
       width="26.458332"
       height="1.8520833"
       x="91.770836"
       y="153.26251" />
    <rect
       style="fill:#606060;fill-opacity:1;stroke-width:0.699999;stroke-linecap:round;stroke-linejoin:round;stop-color:#000000"
       id="rect2138"
       width="26.458332"
       height="1.8520833"
       x="91.770836"
       y="155.11459" />
    <circle
       style="fill:#606060;fill-opacity:1;stroke-width:0.7;stroke-linecap:round;stroke-linejoin:round;stop-color:#000000"
       id="circle1549"
       cx="105"
       cy="155.11459"
       r="4.6302085" />
    <circle
       style="fill:#ffffff;fill-opacity:1;stroke-width:0.7;stroke-linecap:round;stroke-linejoin:round;stop-color:#000000"
       id="path1309"
       cx="105"
       cy="155.11459"
       r="2.6458333" />
    <path
       id="path2222"
       style="fill:#000000;fill-opacity:0.25;stroke-width:2.64567;stroke-linecap:round;stroke-linejoin:round;stop-color:#000000"
       d="M 50 0 A 50.000001 50.000001 0 0 0 0 50 A 50.000001 50.000001 0 0 0 20.810547 90.556641 A 50.000001 50.000001 0 0 1 9.25 58.625 A 50.000001 50.000001 0 0 1 59.25 8.625 A 50.000001 50.000001 0 0 1 88.453125 18.042969 A 50.000001 50.000001 0 0 0 50 0 z "
       transform="matrix(0.26458333,0,0,0.26458333,91.770834,141.88542)" />
  </g>
</svg>
