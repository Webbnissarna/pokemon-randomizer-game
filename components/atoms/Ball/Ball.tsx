import { Box } from "theme-ui";
import { useTheme } from "../../../utils/theme";
import { activeStyle, getSizeStyle, sizeMap } from "./Ball.styles";
import type { BallProps } from "./Ball.types";

export default function Ball({
  size,
  disabled,
  onClick,
}: BallProps): JSX.Element {
  const { theme } = useTheme();

  const stateType = disabled ? "disabled" : "active";
  const onClickFunc = disabled ? undefined : onClick;

  return (
    <Box
      role="button"
      sx={{
        ...getSizeStyle(size),
        ...(!disabled && activeStyle),
      }}
      onClick={onClickFunc}
    >
      <svg
        width={sizeMap[size]}
        height={sizeMap[size]}
        version="1.1"
        viewBox="0 0 26.458 26.458"
        xmlns="http://www.w3.org/2000/svg"
        style={{
          filter: theme.shadows.svgFilter.sm,
        }}
      >
        <g
          transform="translate(-91.771 -141.89)"
          strokeLinecap="round"
          strokeLinejoin="round"
        >
          <path
            d="m105 141.89a13.229 13.229 0 0 0-13.229 13.229h26.458a13.229 13.229 0 0 0-13.229-13.229z"
            fill={theme.colors.ball[stateType].top}
            stopColor="#000000"
            strokeWidth=".7"
          />
          <path
            d="m105 168.34a13.229 13.229 0 0 0 13.229-13.229h-26.458a13.229 13.229 0 0 0 13.229 13.229z"
            fill={theme.colors.ball[stateType].bottom}
            stopColor="#000000"
            strokeWidth=".7"
          />
          <g fill={theme.colors.ball[stateType].middle}>
            <rect
              x="91.771"
              y="153.26"
              width="26.458"
              height="1.8521"
              stopColor="#000000"
              strokeWidth=".7"
            />
            <rect
              x="91.771"
              y="155.11"
              width="26.458"
              height="1.8521"
              stopColor="#000000"
              strokeWidth=".7"
            />
            <circle
              cx="105"
              cy="155.11"
              r="4.6302"
              stopColor="#000000"
              strokeWidth=".7"
            />
          </g>
          <circle
            cx="105"
            cy="155.11"
            r="2.6458"
            fill={theme.colors.ball[stateType].bottom}
            stopColor="#000000"
            strokeWidth=".7"
          />
          <path
            transform="matrix(.26458 0 0 .26458 91.771 141.89)"
            d="m50 0a50 50 0 0 0-50 50 50 50 0 0 0 20.811 40.557 50 50 0 0 1-11.561-31.932 50 50 0 0 1 50-50 50 50 0 0 1 29.203 9.418 50 50 0 0 0-38.453-18.043z"
            fillOpacity=".25"
            stopColor="#000000"
            strokeWidth="2.6457"
          />
        </g>
      </svg>
    </Box>
  );
}
