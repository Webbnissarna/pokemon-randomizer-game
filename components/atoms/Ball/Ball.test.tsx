import { fireEvent, screen } from "@testing-library/react";
import { render } from "../../../.jest/render";
import Ball from ".";

describe("Ball", () => {
  it("renders", () => {
    const { asFragment } = render(<Ball size="sm" />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("calls onClick", () => {
    const callback = jest.fn();

    render(<Ball size="sm" onClick={callback} />);

    fireEvent.click(screen.getByRole("button"));

    expect(callback).toHaveBeenCalledTimes(1);
  });

  it("does not call onClick when disabled", () => {
    const callback = jest.fn();

    render(<Ball size="lg" onClick={callback} disabled={true} />);

    fireEvent.click(screen.getByRole("button"));

    expect(callback).not.toHaveBeenCalled();
  });
});
