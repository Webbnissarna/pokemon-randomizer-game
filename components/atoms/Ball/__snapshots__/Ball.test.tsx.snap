// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ball renders 1`] = `
<DocumentFragment>
  <script>
    (function() { try {
  var mode = localStorage.getItem('theme-ui-color-mode');
  if (!mode) return
  document.documentElement.classList.add('theme-ui-' + mode);
} catch (e) {} })();
  </script>
  <button
    class="css-y8g8p5-Box"
  >
    <svg
      height="40"
      style="filter: drop-shadow(rgba(0, 0, 0, 0.25) 0px 2px 2px);"
      version="1.1"
      viewBox="0 0 26.458 26.458"
      width="40"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        stroke-linecap="round"
        stroke-linejoin="round"
        transform="translate(-91.771 -141.89)"
      >
        <path
          d="m105 141.89a13.229 13.229 0 0 0-13.229 13.229h26.458a13.229 13.229 0 0 0-13.229-13.229z"
          fill="var(--theme-ui-colors-ball-active-top)"
          stop-color="#000000"
          stroke-width=".7"
        />
        <path
          d="m105 168.34a13.229 13.229 0 0 0 13.229-13.229h-26.458a13.229 13.229 0 0 0 13.229 13.229z"
          fill="var(--theme-ui-colors-ball-active-bottom)"
          stop-color="#000000"
          stroke-width=".7"
        />
        <g
          fill="var(--theme-ui-colors-ball-active-middle)"
        >
          <rect
            height="1.8521"
            stop-color="#000000"
            stroke-width=".7"
            width="26.458"
            x="91.771"
            y="153.26"
          />
          <rect
            height="1.8521"
            stop-color="#000000"
            stroke-width=".7"
            width="26.458"
            x="91.771"
            y="155.11"
          />
          <circle
            cx="105"
            cy="155.11"
            r="4.6302"
            stop-color="#000000"
            stroke-width=".7"
          />
        </g>
        <circle
          cx="105"
          cy="155.11"
          fill="var(--theme-ui-colors-ball-active-bottom)"
          r="2.6458"
          stop-color="#000000"
          stroke-width=".7"
        />
        <path
          d="m50 0a50 50 0 0 0-50 50 50 50 0 0 0 20.811 40.557 50 50 0 0 1-11.561-31.932 50 50 0 0 1 50-50 50 50 0 0 1 29.203 9.418 50 50 0 0 0-38.453-18.043z"
          fill-opacity=".25"
          stop-color="#000000"
          stroke-width="2.6457"
          transform="matrix(.26458 0 0 .26458 91.771 141.89)"
        />
      </g>
    </svg>
  </button>
</DocumentFragment>
`;
